rubin_sim/maf/mafContrib/filterPairTGapsMetric.py

```diff
@@ -0,0 +1,175 @@
+        
+        timeCol0 = dataSlice['observationStartMJD'][idx0]
+        timeCol1 = dataSlice['observationStartMJD'][idx1]
+
+        #timeCol0 = timeCol0.reshape((len(timeCol0), 1))
+        #timeCol1 = timeCol1.reshape((len(timeCol1), 1))
+        
+        # calculate time gaps matrix
+        #diffmat = np.subtract(timeCol0, timeCol1.T)
+        
+        if self.allgaps:
+            # collect all time gaps
+            if f0==f1:
+                timeCol0 = timeCol0.reshape((len(timeCol0), 1))
+                timeCol1 = timeCol1.reshape((len(timeCol1), 1))
+
+                diffmat = np.subtract(timeCol0, timeCol1.T)
+                diffmat = np.abs( diffmat )
+                # get only triangle part
+                dt_tri = np.tril(diffmat, -1)
+                dT = dt_tri[dt_tri!=0]    # flatten lower triangle 
+            else:
+                # dT = diffmat.flatten()  
+                dtmax = np.max(self.bins_diff) # time gaps window for measure color
+                dT = []
+                for timeCol in timeCol0:
+                    timeCol_inWindow = timeCol1[(timeCol1>=(timeCol-dtmax)) & (timeCol1<=(timeCol + dtmax))]
+                    
+                    dT.append( np.abs(timeCol_inWindow-timeCol ) )
+                
+                dT = np.concatenate(dT) if len(dT)>0 else np.array(dT)
+        else:
+            # collect only nearest 
+            if f0==f1:
+                # get diagonal ones nearest nonzero, offset=1
+                #dT = np.diagonal(diffmat, offset=1)
+                dT = np.diff(timeCol0)
+            else:
+                timeCol0 = dataSlice['observationStartMJD'][idx0]
+                timeCol1 = dataSlice['observationStartMJD'][idx1]
+
+                timeCol0 = timeCol0.reshape((len(timeCol0), 1))
+                timeCol1 = timeCol1.reshape((len(timeCol1), 1))
+
+                # calculate time gaps matrix
+                diffmat = np.subtract(timeCol0, timeCol1.T)
+                # get tgaps both left and right
+                # keep only negative ones
+                masked_ar = np.ma.masked_where(diffmat>=0, diffmat, )
+                left_ar = np.max(masked_ar, axis=1)
+                dT_left = -left_ar.data[~left_ar.mask]
+                
+                # keep only positive ones
+                masked_ar = np.ma.masked_where(diffmat<=0, diffmat)
+                right_ar = np.min(masked_ar, axis=1)
+                dT_right = right_ar.data[~right_ar.mask]
+                
+                dT = np.concatenate([dT_left.flatten(), dT_right.flatten()])
+        
+        return dT
+    
+    def run(self, dataSlice, slicePoint=None):
+                        
+        # sort the dataSlice in order of time.  
+        dataSlice.sort(order='observationStartMJD')
+        
+        fieldRA = np.mean(dataSlice['fieldRA']) 
+        fieldDec = np.mean(dataSlice['fieldDec'])
+        pixId = radec2pix(nside=16, ra=np.radians(fieldRA), dec=np.radians(fieldDec))
```

Comment:
Hi @xiaolng - I think you actually mean to get these from the slicePoint, not the DataSlice. 
The dataSlice will have a variety of values for fieldRA and fieldDec (because we dither). However, what I think you're trying to evaluate is what are the visits which are relevant for *this point on the sky* -- aka, the slicePoint values. 
If you replace 
```fieldRA = np.mean(dataSlice['fieldRA'])```
with 
```ra = slicePoint['ra'] ```
you will get the point on the sky that you actually mean to evaluate (and avoid any problems that may come from wrap-around at RA=0, where the mean value of a field which has a central pointing at say RA=0.2 and another field with a central pointing of say RA=359.2 would be 180 or so .. and a totally different part of the sky. 

Then again -- you don't even need to calculate epixID, if you are looking for the healpix ID of this point on the sky. You can use slicePoint['sid'] for the 'sliceId'. 

Let me suggest though, that you don't even need that for this metric. The values that are returned by the metric are saved in an array that will match the healpix ids .. so just looking at what index value in the metricValues is being used == this is the healpix id.